#! /bin/sh /usr/share/dpatch/dpatch-run
## 26_support_dns_bootoption.dpatch by Michael Prokop <mika@grml.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Support dns bootoption. Usage examples: dns=8.8.8.8 / dns=8.8.8.8,1.2.3.4

@DPATCH@
--- a/scripts/live
+++ b/scripts/live
@@ -429,7 +429,7 @@
 				echo "search ${DNSDOMAIN}" >> /etc/resolv.conf
 			fi
 
-			for i in ${IPV4DNS0} ${IPV4DNS1} ${IPV4DNS1}
+			for i in ${IPV4DNS0} ${IPV4DNS1} ${IPV4DNS1} ${DNSSERVER1} ${DNSSERVER2}
 			do
 				if [ -n "$i" ] && [ "$i" != 0.0.0.0 ]
 				then
--- a/scripts/live-bottom/23networking_grml
+++ b/scripts/live-bottom/23networking_grml
@@ -44,7 +44,12 @@
 
 # prepare a new /etc/network/interfaces file (and, possibly, a new /etc/resolv.conf)
 IFFILE="/root/etc/network/interfaces"
-RESOLVCONF="/root/etc/resolv.conf"
+if [ -L /root/etc/resolv.conf ] ; then
+  # assume we have resolvconf
+  RESOLVCONF=/root/etc/resolvconf/resolv.conf.d/base
+else
+  RESOLVCONF="/root/etc/resolv.conf"
+fi
 
 # config for loopback networking
 cat > $IFFILE << EOF
@@ -105,3 +110,17 @@
 
     echo>> $IFFILE
 done
+
+# dns bootoption
+if [ -n "$DNSSERVER1" ]
+then
+	# disable any existing entries
+	if [ -r $RESOLVCONF ]
+	then
+		sed -i 's/nameserver/# nameserver/' $RESOLVCONF
+	fi
+	for i in $DNSSERVER1 $DNSSERVER2
+	do
+		echo "nameserver $i" >> $RESOLVCONF
+	done
+fi
--- a/scripts/live-helpers
+++ b/scripts/live-helpers
@@ -59,6 +59,19 @@
 				BOOTIF="${x#BOOTIF=}"
 				;;
 
+			dns=*)
+			  	DNSSERVER="${ARGUMENT#*=}"
+				if echo "${DNSSERVER}" | grep -q , ; then
+					DNSSERVER1="${DNSSERVER%,*}"
+					DNSSERVER2="${DNSSERVER#*,}"
+					export DNSSERVER1 DNSSERVER2
+				else
+					DNSSERVER1="$DNSSERVER"
+					export DNSSERVER1
+				fi
+				unset DNSSERVER
+				;;
+
 			debug)
 				DEBUG="Yes"
 				export DEBUG
